ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -0,0 +1,241 @@
+   * @param logSegmentName the log segment name of the record that is being processed
+   * @param backwardSearchSpan a {@link FileSpan} that spans from the beginning of the index to the beginning of the
+   *                           previous index segment and is used for looking up corresponding put record when applicable
+   * @throws StoreException
+   */
+  private void populateLogSegmentMap(Map<String, Long> logSegmentValidSizeMap, long referenceTimeInMs,
+      MessageInfo messageInfo, long lastModifiedTimeInMs, String logSegmentName, FileSpan backwardSearchSpan)
+      throws StoreException {
+
+    if (messageInfo.isDeleted()) {
+      // delete record
+      updateMapHelper(logSegmentValidSizeMap, logSegmentName, messageInfo.getSize());
+      if (lastModifiedTimeInMs < referenceTimeInMs) {
+        // delete that needs to be accounted for
+        IndexValue putIndexValue = index.findKey(messageInfo.getStoreKey(), backwardSearchSpan);
+        if (putIndexValue != null && !putIndexValue.isFlagSet(IndexValue.Flags.Delete_Index) && !isExpired(
+            putIndexValue.getExpiresAtMs(), referenceTimeInMs)) {
+          // correct previously counted value only if the following conditions are true:
+          // 1. the put and delete record are not in the same index segment (putIndexValue != null)
+          // 2. the found record is actually a put record (!putIndexValue.isFlagSet)
+          // 3. the found put record is not expired
+          updateMapHelper(logSegmentValidSizeMap, putIndexValue.getOffset().getName(), putIndexValue.getSize() * -1);
+        }
+      }
+    } else if (!isExpired(messageInfo.getExpirationTimeInMs(), referenceTimeInMs)) {
+      // put record that is not expired
+      updateMapHelper(logSegmentValidSizeMap, logSegmentName, messageInfo.getSize());
+    }
+  }
+
+  private boolean isExpired(long expirationTimeInMs, long referenceTimeInMs) {
+    return expirationTimeInMs < referenceTimeInMs && expirationTimeInMs != Utils.Infinite_Time;
+  }
+
+  /**
+   * Helper function to update nested map data structure.
+   * @param nestedMap nested {@link Map} to be updated
+   * @param firstKey of the nested map
+   * @param secondKey of the nested map
+   * @param value the value to be added at the corresponding entry
+   */
+  private void updateNestedMapHelper(Map<String, Map<String, Long>> nestedMap, String firstKey, String secondKey,
+      Long value) {
+    if (!nestedMap.containsKey(firstKey)) {
+      nestedMap.put(firstKey, new HashMap<String, Long>());
+    }
+    updateMapHelper(nestedMap.get(firstKey), secondKey, value);
+  }
+
+  /**
+   * Helper function to update map data structure.
+   * @param map {@link Map} to be updated
+   * @param key of the map
+   * @param value the value to be added at the corresponding entry
+   */
+  private void updateMapHelper(Map<String, Long> map, String key, Long value) {
+    Long existingValue = map.get(key);
```

Comment:
you can simplify this. may be remove the method too. 
Long newValue = map.containsKey(key) ?  map.get(key) + value: value;
map.put(key, newValue);